### PR TITLE
Ruby 1.9.3 support for us dinosaurs

### PR DESCRIPTION
**Pull request by astorije**
Do not even think about merging this.
Do not even think about running this in production.

Opening this PR nonetheless to see Travis CI run and give its report. See #148 for more details.

This will be used to run a 24-hour hackathon, and might stay around if folks are willing to experiment and prototype with GraphQL on their outdated projects.

### CODE REVIEW

**Comment by gjtorikian** [comment]
Damn.  Massive respect for this for the hackathon and the sneak-into-the-company tactic.

Good luck!

**Comment by astorije** [comment]
*Live Hacking*

If anyone has an idea how to fix this, that'd be great:

```
/home/travis/build/rmosolgo/graphql-ruby/lib/graphql/define.rb:2:in `require': /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/define/assign_connection.rb:6: syntax error, unexpected '=', expecting ')' (SyntaxError)
...pe_defn, *field_args, options = {}, **field_kwargs, &field_b...
...                               ^
/home/travis/build/rmosolgo/graphql-ruby/lib/graphql/define/assign_connection.rb:9: syntax error, unexpected tPOW, expecting ')'
...call(type_defn, *field_args, **field_kwargs, &field_block)
...                               ^
/home/travis/build/rmosolgo/graphql-ruby/lib/graphql/define/assign_connection.rb:15: syntax error, unexpected keyword_end, expecting $end
```

From my change:

```rb
      ### Ruby 1.9.3 support
      # def self.call(type_defn, *field_args, max_page_size: nil, **field_kwargs, &field_block)
      def self.call(type_defn, *field_args, options = {}, **field_kwargs, &field_block)
        max_page_size = options.fetch(:max_page_size, 'nil')
```

Yeah, it's horrendous, I know, but I'm swimming in unknown territories right now 😅.

**Comment by astorije** [comment]
Alright, calling it a night because right now I feel 🙈 🙊 🚿 💤 😴.

However, I am stuck at the moment, Travis says:

```
/home/travis/build/rmosolgo/graphql-ruby/lib/graphql/schema/reduce_types.rb:63:in `validate_type': Possible type for Beverage is invalid: field "flavors" argument "limit" type must be a valid input type (Scalar or InputObject), not NilClass () (GraphQL::Schema::InvalidTypeError)
	from /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/schema/reduce_types.rb:28:in `reduce_type'
	from /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/schema/reduce_types.rb:50:in `block in crawl_type'
	from /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/schema/reduce_types.rb:49:in `each'
	from /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/schema/reduce_types.rb:49:in `crawl_type'
	from /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/schema/reduce_types.rb:30:in `reduce_type'
	from /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/schema/reduce_types.rb:9:in `block in reduce'
	from /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/schema/reduce_types.rb:8:in `each'
	from /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/schema/reduce_types.rb:8:in `reduce'
	from /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/schema.rb:136:in `types'
	from /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/schema.rb:124:in `define'
	from /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/define/instance_definable.rb:130:in `define'
	from /home/travis/build/rmosolgo/graphql-ruby/spec/support/dairy_app.rb:355:in `<top (required)>'
...
```

The first thing I can find that I edited, going down the stacktrace, is [this](https://github.com/astorije/graphql-ruby/blob/4929478441e5ca1cf55036518f191ae8b05012a3/lib/graphql/schema.rb#L120-L129):

```rb
    ### Ruby 1.9.3 unofficial support
    # def define(**kwargs, &block)
    def define(kwargs = {}, &block)
      super
      types
      # Assert that all necessary configs are present:
      validation_error = Validation.validate(self)
      validation_error && raise(NotImplementedError, validation_error)
      nil
    end
```

@rmosolgo, do you see anything wrong there?
Not really sure what to do now, so I would love some insight.

I feel like I'm not too far, still very hopeful that I don't have to entirely switch project for my hackathon tomorrow night! 😄

**Comment by rmosolgo** [comment]
Amazing!! It looks like you're making great progress. 

The [field in question](https://github.com/rmosolgo/graphql-ruby/blob/master/spec/support/dairy_app.rb#L97) has two things that make it stand out: 

- It has a list type `types[types.String]` 
- The `field` helper is given a block 

Just looking at the diff, I'm not sure how either of those would cause that issue (`field.type` returning nil) but I'm going to pull your branch and try to poke around!

**Comment by rmosolgo** [comment]
My mistake, it was the argument, not the field! 

I was able to make the tests run (but fail) with these commits: https://github.com/astorije/graphql-ruby/compare/astorije/ruby-1.9.3...rmosolgo:rmosolgo/ruby-1.9.3

**Comment by rmosolgo** [comment]
In cases like this :

```
def call(target, name = nil, desc = nil, options = {}); end 
```

Sometimes `name` or `desc` would hold the options hash!

**Comment by astorije** [comment]
> but I'm going to pull your branch and try to poke around!

You are amazing, thank you so much!

I forgot to mention that I enabled edits from maintainers so you could add commits to my branch if you wanted to :)

> Sometimes `name` or `desc` would hold the options hash!

Duh, why didn't I think about that! I guess that's why you shouldn't code (well, that was monkey-coding) until 4am :D

**Comment by rmosolgo** [comment]
Oh sure, I saw you were making annotations on some (all?) changes, so I didn't want to push code on you if it wasn't annotated correctly! 

Also, I had to mess with the Gemspec, my version wasn't finding `Date.today` (it got a no method error). 

If you want those particular commits, feel free to merge: https://github.com/astorije/graphql-ruby/pull/1

**Comment by astorije** [comment]
Yeah, I cherry-picked your commits.

I tried an alternative method, far less elegant than trying to parse the arguments: making sure that optional arguments are actually passed as part of `options` at calling time. So far, not so unlucky.
And searching for `def.*nil.*options` only returns 4 results, so that should be manageable.
I believe I'm just having troubles with [this guy](https://github.com/astorije/graphql-ruby/blob/astorije/ruby-1.9.3/lib/graphql/schema.rb#L143-L146) for now.

To be continued...

**Comment by rmosolgo** [comment]
I think there's an option called `quirks_mode` that allows JSON to dump things other than a hash, but I'm having trouble finding the details on it!

**Comment by rmosolgo** [comment]
I gave it a try: 

```ruby 
~/code/graphql $ ruby -v
ruby 1.9.3p551 (2014-11-13 revision 48407) [x86_64-darwin14.5.0]
~/code/graphql $ irb
require "json"
# => true
JSON.generate(1)
# JSON::GeneratorError: only generation of JSON objects or arrays allowed
# 	from /Users/rmosolgo/.rbenv/versions/1.9.3-p551/lib/ruby/1.9.1/json/common.rb:216:in `generate'
# 	from /Users/rmosolgo/.rbenv/versions/1.9.3-p551/lib/ruby/1.9.1/json/common.rb:216:in `generate'
# 	from (irb):2
# 	from /Users/rmosolgo/.rbenv/versions/1.9.3-p551/bin/irb:12:in `<main>'
JSON.generate(1, quirks_mode: true)
# => "1"
```

Maybe we need some of those in the gem source!

**Comment by astorije** [comment]
Cool, I moved from `dump` to `generate` with that option (`dump` doesn't have an option like that, but it looks like it should enable `quirks_mode` by default, I'm not too sure... but tests pass so, meh!).

**Comment by astorije** [comment]
I am a bit confused by this:

```rb
DummySchema.execute(%|
      {
        __type(name: "Cheese") {
          ...
        }
      }
    |)
```

calls:

```rb
    def execute(*args)
      query_obj = GraphQL::Query.new(self, *args)
      query_obj.result
    end
```

which then calls:

```rb
def initialize(schema, options = {})
      query_string = options.fetch(:query_string, nil)
```

which generates:

```rb
NoMethodError:         NoMethodError: undefined method `fetch' for #<String:0x00000003f0d658>
            /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/query.rb:35:in `initialize'
            /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/schema.rb:144:in `new'
            /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/schema.rb:144:in `execute'
            /home/travis/build/rmosolgo/graphql-ruby/spec/graphql/introspection/input_value_type_spec.rb:44:in `block (2 levels) in <top (required)>'
```

This is also related:

```rb
ArgumentError:         ArgumentError: wrong number of arguments (3 for 2)
            /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/query.rb:34:in `initialize'
            /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/schema.rb:144:in `new'
            /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/schema.rb:144:in `execute'
            /home/travis/build/rmosolgo/graphql-ruby/spec/graphql/query/arguments_spec.rb:77:in `block (3 levels) in <top (required)>'
```

I feel like I'm missing something obvious...

---

BTW, I must say, I **tremendously** appreciate both your help and the massive test suite you have built. This is amazing.

**Comment by astorije** [comment]
Also found out that:

```rb
    ### Ruby 1.9.3 unofficial support
    # def define(**kwargs, &block)
    def define(kwargs = {}, &block)
      super
      types
      # Assert that all necessary configs are present:
      validation_error = Validation.validate(self)
      validation_error && raise(NotImplementedError, validation_error)
      nil
    end
```

is most likely to be the part causing:

```rb
  test_0005_sets correct default values for complex field argumentsERROR (0.00s)
GraphQL::Schema::InvalidTypeError:         GraphQL::Schema::InvalidTypeError: Query is invalid: field "post" argument "varied" default value {:id=>"123", :int=>234, :float=>2.3, :enum=>:foo, :sub=>[{:string=>"str"}]} is not valid for type Varied (key not found: :foo)
            /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/schema/reduce_types.rb:63:in `validate_type'
            /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/schema/reduce_types.rb:28:in `reduce_type'
            /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/schema/reduce_types.rb:9:in `block in reduce'
            /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/schema/reduce_types.rb:8:in `each'
            /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/schema/reduce_types.rb:8:in `reduce'
            /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/schema.rb:136:in `types'
            /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/schema.rb:124:in `define'
            /home/travis/build/rmosolgo/graphql-ruby/lib/graphql/define/instance_definable.rb:130:in `define'
            /home/travis/build/rmosolgo/graphql-ruby/spec/graphql/schema/loader_spec.rb:109:in `block (2 levels) in <top (required)>'
````

Looks like there is something wrong with how I dealt with double splats, but I can't figure out what.

Aaargh, only 18 errors left...

**Comment by rmosolgo** [comment]
When it comes to `Schema#execute`, I think you have to massage the arguments a bit, for example: 

```ruby
def execute(*args)
  # Check for provided options, or create some new ones: 
  options = args.last.is_a?(Hash) ? args.last : {}

  if args.first.is_a?(String) # it's a query string 
    # Move the positional argument into `options`: 
    options[:query_string] = args.first
  end 
  Query.new(self, options).result 
end 
```

**Comment by rmosolgo** [comment]
I think that will also fix the `ArgumentError: wrong number of arguments (3 for 2)` because: 

- __previously__ it would be called with `(schema, query_string, options)`
- __after that change__ it would be called with `(schema, options)`

**Comment by rmosolgo** [comment]
Pushed a commit with a change like that one and one more that looked similar

**Comment by rmosolgo** [comment]
it's working!

**Comment by astorije** [comment]
Whaaaaaat, all tests pass!! 🎉 🙌 🍾 🎆 

You sir are seriously the best, thanks a ton!

**Comment by rmosolgo** [comment]
Happy to help out, thanks for leading the charge on this! 

If it catches on at the hackathon, I'm happy to talk about options for maintaining a `1.9.3` branch. (But I can't merge it into master ... keyword args are just too good to give up!)

**Comment by astorije** [comment]
> I'm happy to talk about options for maintaining a 1.9.3 branch. (But I can't merge it into master ... keyword args are just too good to give up!)

Cool! I do believe 1 or 2 changes made to this branch should be applied to `master`, but yeah, no, I never intended this to go to master! :-)
Added a mention in the README of this branch in case some crazy person would run this as-is.

**Comment by rmosolgo** [comment]
Well .... how'd it go?!

**Comment by astorije** [comment]
@rmosolgo, it went actually really well! I'm in love with GraphQL, thank you so much for your efforts with this project and this PR!

Our execs and leaders also liked it a lot and asked why we weren't doing this already for our API lol, so it might become a reality. My guess is that if it does, it might happen before or while we transition to Ruby 2 (it is a really massive project across many teams, hundreds of servers, etc., we cannot transition in one day :/), so that would be internal and experimental until transition is over. In such situation, I would maintain the 1.9.3 version, while keeping a note that it should never be used in production.

I learned a few lessons from this hackathon regarding this branch: some changes should be moved to master, but most should obviously not. More importantly, my temporary hack of moving `query_string`, while nice for the API I think, isn't compatible with official gem so anyone using this branch could not just reuse their code after upgrading to Ruby 2.
What are your thought on making `query_string` a part of the keyword arguments in https://github.com/rmosolgo/graphql-ruby/pull/306/files#diff-5cdc567175a410efb6ad535ca2d69939L32 ? Obviously, only if their is a way to *not* break current API.

I'll be editing that PR and getting back as I make progress. What are your thoughts on maintaining a branch for Ruby 1.9.3?

Also, note that I had to make `graphiql-rails` compatible with Ruby 1.9.3 and Rails 3.2.22.1 as well to have a nice demo for the hackathon. I didn't open a PR because tests were not passing at all (but project worked really well still) and, well, I was kinda busy 😅 But it's [here](https://github.com/rmosolgo/graphiql-rails/compare/master...astorije:astorije/ruby-1.9.3) and I'll get to that after this current branch is ready.

**Comment by astorije** [comment]
@rmosolgo, I opened a few PRs extracted from this PR for things that I believe should be on master, even though they pass there and specifically not on Ruby 1.9.3.
I will clean up that PR after you merge them, if you are happy with them obviously.

**Comment by rmosolgo** [comment]
:tada: glad to hear it went well! It's fun to watch people have the "aha!" moment for GraphQL, isn't it :) 

Personally, I don't want to maintain a 1.9.3 branch, but I'd be happy to keep working with you to maintain one (like we did here), for example, upstreaming some improvements, keeping a compatible test suite, and pitching in on code changes. 


>  making `query_string` a part of the keyword arguments

Let me see, the difference is ... 

```ruby 
# from ...
Schema.execute(query_string)
# to ... 
Schema.execute(query_string: query_string)
```

Can you help me understand the advantage of the second one? On the one hand, I understand the readability value of _named_ arguments, but on the other hand, I think of the string as the "main" input to GraphQL execution, so it makes sense to me for it to be the first, required argument. 

Is there another angle which I've overlooked? 

One more thing, I'd like to get the `JSON`-related changes in `master`. Recently, someone else reported a bug for that, where another JSON implementation was :boom:-ing. Do you have time to PR that? Or do you mind if I grab those changes?

**Comment by astorije** [comment]
> I don't want to maintain a 1.9.3 branch, but I'd be happy to keep working with you to maintain one (like we did here)

Absolutely, I was not asking you to maintain one :)

> Is there another angle which I've overlooked?

Not really, just consistency between:

```rb
Schema.execute(query_string: query_string)
# And
Schema.execute(query_string: query_string, variables: variables)
```

And every other optional arguments in http://www.rubydoc.info/github/rmosolgo/graphql-ruby/GraphQL/Query#initialize-instance_method.
Living in a post-apocalyptic Ruby 1.9.3 world, I am not accustomed to keyword arguments yet. What is the rationale when you defined an optional argument, between making it `foo = bar` and `foo: bar`?

> Do you have time to PR that?

Sure thing!

File: graphql.gemspec [hunk 1]
```diff
@@ -5,7 +5,7 @@ require "graphql/version"
 Gem::Specification.new do |s|
   s.name        = "graphql"
   s.version     = GraphQL::VERSION
-  s.date        = Date.today.to_s
+  s.date        = "2016-10-12"
```
Comment (astorije) [inline graphql.gemspec]: @rmosolgo, I'm not sure why you needed this. I never got this issue, [and I just tried without](https://github.com/rmosolgo/graphql-ruby/pull/306/commits/0d5e1a7a6f2647c7c19296a672c829d7e27e10e2), looks like [it built fine](https://travis-ci.org/rmosolgo/graphql-ruby/builds/168083322)...
Did you see that locally or on Travis CI?

**Comment by rmosolgo** [comment]
> `foo = bar` and `foo: bar`

the risk for positional arguments is just how they can grow and grow... what starts innocently as a method with one parameter can become a monstrosity of nameless, arbitrary positional arguments. 

Here's an [example API](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/drawImage):

image

When you have that long list of args, you have to "fake it" for keywords with comments, eg 

image

So for me, more than two positional arguments is a bit of a red flag. Maybe it's ok if the scope of the code is small (private API, called in the same file, etc) but otherwise ... yikes! (hehe, don't search graphql-ruby for that rule ... I've broken it many times 😬 )

**Comment by rmosolgo** [comment]
I'm glad this experiment basically worked! This branch has fallen far behind and I don't know of anyone actively using it, so I'm going to close the PR. Thanks again for the improvements you upstreamed from this work. If anyone _is_ maintaining a 1.9.x branch, I'd love to link to it from the documentation!

**Pull request closed without merging**